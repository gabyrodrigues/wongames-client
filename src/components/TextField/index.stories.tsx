import React from "react";
import { StoryObj, Meta } from "@storybook/react";
import TextField, { TextFieldProps } from ".";
import { Email } from "@styled-icons/material-outlined";

export default {
  title: "TextField",
  component: TextField,
  args: {
    label: "E-mail",
    labelFor: "Email",
    id: "Email",
    initialValue: "",
    placeholder: "john.cage@gmail.com"
  },
  argTypes: {
    onChange: { action: "changed" }
  }
} as Meta<TextFieldProps>;

export const Default: StoryObj = {
  decorators: [
    (Story) => (
      <div style={{ maxWidth: "300px", padding: "15px" }}>
        <Story />
      </div>
    )
  ]
};

export const WithIcon: StoryObj = {
  args: {
    icon: <Email />,
    iconPosition: "left"
  },
  argTypes: {
    icon: { type: "" }
  },
  decorators: [
    (Story) => (
      <div style={{ maxWidth: "300px", padding: "15px" }}>
        <Story />
      </div>
    )
  ]
};

export const WithError: StoryObj = {
  args: {
    error: "Ops...something is wrong!",
    icon: <Email />
  },
  argTypes: {
    icon: { type: "" }
  },
  decorators: [
    (Story) => (
      <div style={{ maxWidth: "300px", padding: "15px" }}>
        <Story />
      </div>
    )
  ]
};
