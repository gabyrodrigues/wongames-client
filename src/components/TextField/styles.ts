import styled, { css } from "styled-components";

export const InputWrapper = styled.div`
  ${({ theme }) => css`
    background: ${theme.colors.lightGray};
    border-radius: 0.2rem;
    border: 0.2rem solid;
    border-color: ${theme.colors.lightGray};
    display: flex;
    padding: 0 ${theme.spacings.xsmall};

    &:focus-within {
      box-shadow: 0 0 0.5rem ${theme.colors.primary};
    }
  `}
`;

export const Input = styled.input`
  ${({ theme }) => css`
    background: transparent;
    border: 0;
    color: ${theme.colors.black};
    font-family: ${theme.font.family};
    font-size: ${theme.font.sizes.medium};
    outline: none;
    padding: ${theme.spacings.xxsmall};
    width: 100%;
  `}
`;

export const Label = styled.label`
  ${({ theme }) => css`
    font-size: ${theme.font.sizes.small};
    color: ${theme.colors.black};
    cursor: pointer;
  `}
`;

export const Wrapper = styled.div``;

export const Icon = styled.div`
  ${({ theme }) => css`
    color: ${theme.colors.gray};
    display: flex;
    width: 2.2rem;

    & > svg {
      width: 100%;
    }
  `}
`;
