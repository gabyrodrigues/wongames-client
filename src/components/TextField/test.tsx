import { screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { renderWithTheme } from "@utils/tests/helpers";
import TextField from ".";
import { Email } from "@styled-icons/material-outlined";

describe("<TextField />", () => {
  it("Renders with Label", () => {
    renderWithTheme(
      <TextField
        label="Label"
        labelFor="Field"
        id="Field"
      />
    );

    expect(screen.getByLabelText("Label")).toBeInTheDocument();
  });

  it("Renders without Label", () => {
    renderWithTheme(<TextField />);

    expect(screen.queryByLabelText("Label")).not.toBeInTheDocument();
  });

  it("Renders with placeholder", () => {
    renderWithTheme(<TextField placeholder="hey you" />);

    expect(screen.getByPlaceholderText("hey you")).toBeInTheDocument();
  });

  it("Changes its value when typing", async () => {
    const onChange = jest.fn();
    renderWithTheme(
      <TextField
        onChange={onChange}
        label="TextField"
        labelFor="TextField"
        id="TextField"
      />
    );

    const input = screen.getByRole("textbox");
    const text = "This is my new text";

    userEvent.type(input, text);

    await waitFor(() => {
      expect(input).toHaveValue(text);
      expect(onChange).toHaveBeenCalledTimes(text.length);
    });
  });

  it("Is accessible by tab", async () => {
    renderWithTheme(
      <TextField
        label="TextField"
        labelFor="TextField"
        id="TextField"
      />
    );

    const input = screen.getByLabelText("TextField");
    expect(document.body).toHaveFocus();

    await waitFor(() => {
      userEvent.tab();
      expect(input).toHaveFocus();
    });
  });
  it("should render an icon version", () => {
    renderWithTheme(
      <TextField
        label="TextField"
        labelFor="TextField"
        id="TextField"
        icon={<Email data-testid="icon" />}
      />
    );

    expect(screen.getByTestId("icon")).toBeInTheDocument();
  });
  it("should render an icon in the right position", () => {
    renderWithTheme(
      <TextField
        label="TextField"
        labelFor="TextField"
        id="TextField"
        icon={<Email data-testid="icon" />}
        iconPosition="right"
      />
    );

    expect(screen.getByTestId("icon").parentElement).toHaveStyle({ order: 1 });
  });
});
