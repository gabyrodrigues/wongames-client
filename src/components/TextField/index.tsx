import React, { InputHTMLAttributes, ReactNode, useState } from "react";
import * as S from "./styles";

export interface TextFieldProps extends InputHTMLAttributes<HTMLInputElement> {
  label?: string;
  labelFor?: string;
  initialValue?: string;
  icon?: ReactNode;
  iconPosition?: "left" | "right";
  disabled?: boolean;
  error?: string;
}

export default function TextField({
  onChange,
  label,
  labelFor = "",
  initialValue = "",
  icon,
  iconPosition = "left",
  disabled = false,
  error,
  ...props
}: TextFieldProps) {
  const [value, setValue] = useState(initialValue);

  function handleOnChange(event: React.ChangeEvent<HTMLInputElement>) {
    const newValue = event.currentTarget.value;
    setValue(newValue);

    !!onChange && onChange(event);
  }
  return (
    <S.Wrapper
      disabled={disabled}
      error={!!error}>
      {!!label && <S.Label htmlFor={labelFor}>{label}</S.Label>}
      <S.InputWrapper>
        {!!icon && <S.Icon $iconPosition={iconPosition}>{icon}</S.Icon>}
        <S.Input
          type="text"
          onChange={handleOnChange}
          value={value}
          $iconPosition={iconPosition}
          {...props}
          disabled={disabled}
        />
      </S.InputWrapper>
      {!!error && <S.Error>{error}</S.Error>}
    </S.Wrapper>
  );
}
