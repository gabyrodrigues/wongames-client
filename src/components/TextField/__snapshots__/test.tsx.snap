// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextField /> Renders with error 1`] = `
.c4 {
  background: #EAEAEA;
  border-radius: 0.2rem;
  border: 0.2rem solid;
  border-color: #EAEAEA;
  display: flex;
  padding: 0 1.6rem;
}

.c4:focus-within {
  box-shadow: 0 0 0.5rem #F231A5;
}

.c8 {
  background: transparent;
  border: 0;
  color: #030517;
  font-family: Poppins,-apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Oxygen,Ubuntu,Cantarell,'Open Sans','Helvetica Neue',sans-serif;
  font-size: 1.6rem;
  outline: none;
  padding: 0.8rem 0;
  padding-left: 1.6rem;
  width: 100%;
}

.c2 {
  font-size: 1.4rem;
  color: #030517;
  cursor: pointer;
}

.c0 .c3 {
  border-color: #FF6347;
}

.c0 .c5,
.c0 .c1 {
  color: #FF6347;
}

.c6 {
  color: #8F8F8F;
  display: flex;
  order: 0;
  width: 2.2rem;
}

.c6>svg {
  width: 100%;
}

.c9 {
  color: #FF6347;
  font-size: 1.2rem;
}

.c7 {
  display: inline-block;
  vertical-align: middle;
  overflow: hidden;
}

<div
  class="c0"
>
  <label
    class="c1 c2"
    for="TextField"
  >
    TextField
  </label>
  <div
    class="c3 c4"
  >
    <div
      class="c5 c6"
    >
      <svg
        aria-hidden="true"
        class="c7"
        data-testid="icon"
        fill="currentColor"
        focusable="false"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M22 6c0-1.1-.9-2-2-2H4c-1.1 0-2 .9-2 2v12c0 1.1.9 2 2 2h16c1.1 0 2-.9 2-2V6zm-2 0-8 5-8-5h16zm0 12H4V8l8 5 8-5v10z"
        />
      </svg>
    </div>
    <input
      class="c8"
      type="text"
      value=""
    />
  </div>
  <p
    class="c9"
  >
    Error message
  </p>
</div>
`;
